### PR TITLE
ARROW-6666: [Rust] Datafusion parquet string literal support

### PR DESCRIPTION
**Pull request by maxburke**
This change also required having to add comparison predicates for UTF8 strings to Arrow's comparison kernels. Included also are support for the LIKE and NOT LIKE operators.

### CODE REVIEW

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -22,17 +22,50 @@
...
 use crate::error::{ArrowError, Result};
 
 /// Helper function to perform boolean lambda function on values from two arrays, this
 /// version does not attempt to use SIMD.
-pub fn compare_op<T, F>(
+macro_rules! compare_op {
+    ($left: expr, $right:expr, $op:expr) => {{
+        if $left.len() != $right.len() {
+            return Err(ArrowError::ComputeError(
+                "Cannot perform math operation on arrays of different length".to_string(),
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/comparison.rs]: This error message refers to "math operation" which seems incorrect?

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -41,6 +74,79 @@ where
...
+            map: HashMap::new(),
+        }
+    }
+
+    fn match_str(&'a mut self, a: &str, pat: &'a str) -> bool {
+        let re = if let Some(ref regex) = self.map.get(pat) {
+            regex
+        } else {
+            let re_pattern = pat.replace("%", ".*").replace("_", ".");
+            let re = Regex::new(&re_pattern).unwrap();
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/comparison.rs]: Can we avoid the `unwrap` here? Or at least use `expect` with an appropriate error message?

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 3]
```diff
@@ -41,6 +74,79 @@ where
...
+            self.map.insert(pat, re);
+
+            self.map.get(pat).unwrap()
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/comparison.rs]: Another `unwrap`

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 4]
```diff
@@ -41,6 +74,79 @@ where
...
+
+        re.is_match(a)
+    }
+}
+
+pub fn like_utf8(left: &StringArray, right: &StringArray) -> Result<BooleanArray> {
+    let mut map = HashMap::new();
+    if left.len() != right.len() {
+        return Err(ArrowError::ComputeError(
+            "Cannot perform math operation on arrays of different length".to_string(),
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/comparison.rs]: Error message refers to "math" again

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 5]
```diff
@@ -41,6 +74,79 @@ where
...
+
+    let mut result = BooleanBufferBuilder::new(left.len());
+    for i in 0..left.len() {
+        let haystack = left.value(i);
+        let pat = right.value(i);
+        let re = if let Some(ref regex) = map.get(pat) {
+            regex
+        } else {
+            let re_pattern = pat.replace("%", ".*").replace("_", ".");
+            let re = Regex::new(&re_pattern).unwrap();
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/comparison.rs]: `unwrap`

**Review by andygrove — COMMENTED** [review]
This looks great! I left some comments about the calls to `unwrap` and some error messages that are incorrect but otherwise LGTM.

File: rust/arrow/src/compute/kernels/comparison.rs (hunk 1 above)
Comment (maxburke) [reply rust/arrow/src/compute/kernels/comparison.rs]: Looks like the original comparisons.rs was referring to "math operation" as well: 

https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/comparison.rs#L46

But I can definitely fix that!

**Comment by maxburke** [comment]
Done!

**Review by andygrove — APPROVED** [review]
LGTM!

**Pull request closed without merging**